### PR TITLE
Add domain objects and proto mappers

### PR DESCRIPTION
**Pull request by volkangurel**
Add domain objects and proto mappers to prepare for the executable entrypoints refactor

### CODE REVIEW

File: layer/clients/model_catalog.py [hunk 1]
```diff
@@ -1,56 +1,54 @@
...
 from layer.contracts.project_full_name import ProjectFullName
 from layer.contracts.tracker import ResourceTransferState
-from layer.exceptions.exceptions import LayerClientException
+from layer.exceptions.exceptions import (
+    LayerClientException,
+    UnexpectedModelTypeException,
+)
 from layer.flavors.base import ModelRuntimeObjects
-from layer.flavors.utils import get_flavor_for_proto
+from layer.flavors.utils import get_flavor_for_model, get_flavor_for_proto
```
Comment (cihati) [inline layer/clients/model_catalog.py:46]: not sure if you want to colocate, but both these methods are only used here now.

File: layer/clients/data_catalog.py [hunk 2]
```diff
@@ -152,72 +142,14 @@ def store_dataset(
                     writer.write_batch(chunk)
                     total_rows += chunk.num_rows
```
Comment (cihati) [inline layer/clients/data_catalog.py:143]: why do we need `total_rows` any more?

File: layer/clients/model_catalog.py [hunk 3]
```diff
@@ -218,64 +175,54 @@ def _load_model_runtime_objects(
...
         self,
-        model: Model,
         model_object: ModelObject,
+        train_id: uuid.UUID,
         transfer_state: ResourceTransferState,
     ) -> ModelObject:
-        self._logger.debug(f"Storing given model {model_object} for {model.path}")
+        flavor = get_flavor_for_model(model_object)
+        if flavor is None:
+            raise UnexpectedModelTypeException(type(model_object))
```
Comment (cihati) [inline layer/clients/model_catalog.py:184]: somewhat unrelated, but probably you wanna improve the message `UnexpectedModelTypeException` throws.

**Review by cihati — COMMENTED** [review]
how about these:

```
layer/logged_data/data_logging_request.py:from layerapi.api.service.logged_data.logged_data_api_pb2 import LogDataResponse
layer/logged_data/log_data_runner.py:from layerapi.api.value.logged_data_type_pb2 import LoggedDataType
layer/projects/execution_planner.py:from layerapi.api.entity.operations_pb2 import (
layer/projects/execution_planner.py:from layerapi.api.entity.task_pb2 import Task
layer/projects/execution_planner.py:from layerapi.api.value.language_version_pb2 import LanguageVersion
layer/projects/progress_tracker_updater.py:from layerapi.api.entity.history_event_pb2 import HistoryEvent
layer/projects/progress_tracker_updater.py:from layerapi.api.entity.run_metadata_pb2 import RunMetadata
layer/projects/progress_tracker_updater.py:from layerapi.api.entity.run_pb2 import Run as PBRun
layer/projects/progress_tracker_updater.py:from layerapi.api.entity.task_pb2 import Task as PBTask
layer/training/train.py:from layerapi.api.ids_pb2 import ModelTrainId
layer/training/train.py:from layerapi.api.value.model_flavor_pb2 import ModelFlavor
layer/user_logs.py:from layerapi.api.entity.user_log_line_pb2 import UserLogLine as PBUserLogLine
```

File: layer/training/train.py [hunk 4]
```diff
@@ -121,5 +98,4 @@ def __exit__(
         # the execution of this method. We'd only proceed if there are no exceptions.
         # In happy path, all three parameters would be None.
         # https://docs.python.org/3/reference/datamodel.html#object.__exit__
-        if exception_type is None:
-            self.__complete_train()
+        pass
```
[the remark this answers is not in the document]
Comment (volkangurel) [reply layer/training/train.py:101]: This class is being removed in https://github.com/layerai/sdk/pull/410

File: layer/clients/model_catalog.py (hunk 1 above)
Comment (volkangurel) [reply layer/clients/model_catalog.py:46]: I think let's leave them as is for now, and we'll likely need to move them around when we build artifacts

File: layer/clients/data_catalog.py (hunk 2 above)
Comment (volkangurel) [reply layer/clients/data_catalog.py:143]: we don't, will remove!

File: layer/clients/model_catalog.py (hunk 3 above)
Comment (volkangurel) [reply layer/clients/model_catalog.py:184]: we will redo artifacts soon, let's tackle that then?

**Review by cihati — APPROVED** [review]

**Pull request merged**